flaskbb/plugins/__init__.py

```diff
@@ -8,18 +8,107 @@
     :copyright: (c) 2014 by the FlaskBB Team.
     :license: BSD, see LICENSE for more details.
 """
+import contextlib
+
+import copy
+import os
 from flask import current_app
+from flask import g
+from flask_migrate import upgrade, downgrade, migrate
 from flask_plugins import Plugin
+from flaskbb.extensions import db, migrate as mig
+from flaskbb.management.models import SettingsGroup,Setting
 
-from flaskbb.management.models import SettingsGroup
+@contextlib.contextmanager  # TODO: Add tests
+def plugin_name_migrate(name):
+    """Migrations in this with block will only apply to models from the named plugin"""
+    g.plugin_name = name
+    yield
+    del g.plugin_name
 
 
-class FlaskBBPlugin(Plugin):
+def db_for_plugin(plugin_name,sqla_instance=None):
+    """Labels models as belonging to this plugin.
+    sqla_instance is a valid Flask-SQLAlchemy instance, if None, then the default db is used
+
+    Usage:
+        from flaskbb.plugins import db_for_plugin
+
+        db=db_for_plugin(__name__)
+
+        mytable = db.Table(...)
+
+        class MyModel(db.Model):
+            ...
+
+        """
+    sqla_instance=sqla_instance or db
+    new_db = copy.copy(sqla_instance)
+
+    def Table(*args, **kwargs):
+        new_table = sqla_instance.Table(*args, **kwargs)
+        new_table._plugin = plugin_name
+        return new_table
+
+    new_db.Table = Table
+    return new_db
 
+
+@mig.configure
+def config_migrate(config):
+    """Configuration callback for plugins environment"""
+    plugins = current_app.extensions['plugin_manager'].all_plugins.values()
+    migration_dirs = [p.get_migration_version_dir() for p in plugins]
+    if config.get_main_option('version_table') == 'plugins':
+        config.set_main_option('version_locations', ' '.join(migration_dirs))
+        print config.get_main_option('version_locations')
+    return config
+
+
+class FlaskBBPlugin(Plugin):
     #: This is the :class:`SettingsGroup` key - if your the plugin needs to
     #: install additional things you must set it, else it won't install
     #: anything.
     settings_key = None
+    requires_install=None
+
+    def resource_filename(self, *names):
+        "Returns an absolute filename for a plugin resource."
+        if len(names)==1 and '/' in names[0]:
+            names=names[0].split('/')
+        fname= os.path.join(self.path, *names)
+        if ' ' in fname and not '"' in fname and not "'" in fname:
+            fname='"%s"'%fname
+        return fname
+
+    def get_migration_version_dir(self):
```

Comment:
Same for this. Instead you have always used ``os.path.join(self.path,'migration_versions')``